qbraid/runtime/azure/provider.py

```diff
@@ -0,0 +1,494 @@
+from qbraid_core._import import LazyLoader
+from qbraid_core.sessions import Session
+
+from qbraid.programs.spec import ProgramSpec
+from qbraid.runtime.enums import DeviceType
+from qbraid.runtime.profile import TargetProfile
+from qbraid.runtime.provider import QuantumProvider
+
+from .device import AzureQuantumDevice
+
+pyquil = LazyLoader("pyquil", globals(), "pyquil")
+
+
+class AzureSession(Session):
+    """Azure session class."""
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        tenant_id: str,
+        location_name: str,
+        subscription_id: str,
+        resource_group_name: str,
+        workspace_name: str,
+        storage_account: str,
+        connection_string: str,
+    ):  # pylint: disable= too-many-arguments
+
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.tenant_id = tenant_id
+        self.location_name = location_name
+        self.subscription_id = subscription_id
+        self.resource_group_name = resource_group_name
+        self.workspace_name = workspace_name
+        self.storage_account = storage_account
+        self.connection_string = connection_string
+        self.backends = {
+            "Syntax Checker": "quantinuum.sim.h1-1sc",
+            "Emulator": "quantinuum.sim.h1-1e",
+            "QPU": "quantinuum.qpu.h1-1",
+        }  # will need to add for pyquil
+        self.formats = {"quantinuum": ["honeywell.openqasm.v1", "honeywell.quantum-results.v1"]}
+
+        self.jobs = {}
+        self.session = Session()
+        self._helpers = AzureHelperFunctions()
+
+        self.base_url = (f"https://{location_name}.quantum.azure.com",)
+
+        super().__init__(base_url=self.base_url)
+
+    def get_devices(self) -> list[str, dict[str, Any]]:
+        """Get all Azure Quantum devices."""
+        quantum_access_token = self._helpers.quantum_access_token(
+            self.client_id, self.client_secret, self.tenant_id, self.session
+        )
+
+        url = (
+            f"https://{self.location_name}.quantum.azure.com/subscriptions/{self.subscription_id}"
+            f"/resourceGroups/{self.resource_group_name}/providers/Microsoft.Quantum/"
+            f"workspaces/{self.workspace_name}/providerStatus?api-version=2022-09-12-preview"
+        )
+
+        auth_headers = {"Authorization": f"Bearer {quantum_access_token}"}
+        r = self.session.get(url, headers=auth_headers)
+        devices = r.json()
+
+        devices_dict = self._helpers.device_manager(devices)
+        devices_list = [{"id": k, **v} for k, v in devices_dict.items()]
+
+        return devices_list
+
+    def get_device(self, device_id: str) -> dict[str, Any]:
+        """Get a specific Azure Quantum device."""
+        devices = self.get_devices()
+        device = None
+
+        for i in devices:
+            if i.get("id") == device_id:
+                device = i
+
+        return device
```

Comment:
If ID matches, return right away. If you get through the entire list and no IDs matched, raise a `qbraid.runtime.ResourceNotFoundError`, or similar.

Returning `None` when resource is not found will just figuratively "kick the rock down the road" as to when the exception is raised.